### PR TITLE
Fixed #33616 -- Allowed registering callbacks that can fail in transaction.on_commit().

### PR DESCRIPTION
**Pull request by SirAbhi13**
[Ticket #33616](https://code.djangoproject.com/ticket/33616)

Added on_commit_robust function in the same style as send_robust in signals.

### CODE REVIEW

File: django/db/backends/base/base.py
Comment (David-Wobrock) [inline django/db/backends/base/base.py:746]: When we register the function to run robustly on commit in an atomic bloc, should we also try/catch it when it is actually run?
Meaning that in `db.backends.base.base.BaseDatabaseWrapper.run_and_clear_commit_hooks` we should take into account this info

File: django/db/backends/base/base.py
Comment (David-Wobrock) [inline django/db/backends/base/base.py:753]: Use a similar logging to what we do for robust signals?
See `django.dispatch.dispatcher.Signal.send_robust` and
```
            try:
                response = receiver(signal=self, sender=sender, **named)
            except Exception as err:
                logger.error(
                    "Error calling %s in Signal.send_robust() (%s)",
                    receiver.__qualname__,
                    err,
                    exc_info=err,
                )
```
Something like
```suggestion
                logger.error("Error calling {func.__qualname__} on_commit() ({e}).", exc_info=True)
```

**Review by David-Wobrock — COMMENTED** [review]
Thanks!
We should document this method and add a release note entry in `django/docs/releases/4.2.txt` :)

EDIT: and adding a few unit tests for be worthwhile I guess 👍

**Comment by SirAbhi13** [comment]
Thank you! I will get on it.

I did try to find the Tests of on_commit but was unable to find it, would you please point me to the appropriate file for it's test.

File: django/db/backends/base/base.py
Comment (SirAbhi13) [reply django/db/backends/base/base.py:753]: I will make the appropriate changes.

File: django/db/backends/base/base.py
Comment (SirAbhi13) [reply django/db/backends/base/base.py:746]: Can you please elaborate more of what you expect here?

Comment (David-Wobrock) [reply django/db/backends/base/base.py:746]: When calling `on_commit` there are basically two modes:
- there is no transaction in progress, so we execute the function right away
- we are in an atomic block, so we register the function to execute it later (`self.run_on_commit.append(`)

The first case is handled by the PR, but not the second one. And I'd think that we would need to handle a robust execution in the second case too.
Does that make it clearer? :)

**Comment by David-Wobrock** [comment]
> Thank you! I will get on it.
> 
> I did try to find the Tests of on_commit but was unable to find it, would you please point me to the appropriate file for it's test.

I think they should be in `tests.transaction_hooks.tests.TestConnectionOnCommit`

File: django/db/backends/base/base.py
Comment (SirAbhi13) [reply django/db/backends/base/base.py:746]: Yes! Thanks for the reply. I will commit the changes soon.

**Comment by SirAbhi13** [comment]
@David-Wobrock I have made the asked changes and added tests. Please review.

**Comment by SirAbhi13** [comment]
@felixxm sorry for the ping, i just wanted a review on this PR as it would seem @David-Wobrock is occupied right now.

**Comment by SirAbhi13** [comment]
> > @felixxm sorry for the ping, i just wanted a review on this PR as it would seem @David-Wobrock is occupied right now.
> 
> I didn't forget about the PR, I'll come back to it soon enough :) From my understanding, there is no hurry here since this is a new feature and the team is currently busy with the 4.1 beta release.
> 
> By the way, it can help to set the ticket into a state that suggests that it is ready for review. For instance on https://code.djangoproject.com/ticket/33616, you could unset _Needs tests_ and _Needs documentation_ if you think the PR is fine on this side.

Thanks for the reply. Sorry for the unnecessary ping😔. Won't happen again in the future! Was just excited as it was my first PR in the organization 😓. I will review the changes from my side and leave it up to you for review whenever the maintainers are free.

Thanks again!

File: django/db/backends/base/base.py
Comment (David-Wobrock) [inline django/db/backends/base/base.py:766]:
```suggestion
                    f"Error calling {func.__qualname__} on_commit_robust() ({e}).",
```

File: django/db/backends/base/base.py
Comment (David-Wobrock) [inline django/db/backends/base/base.py:788]: This implies that the `on_commit_robust` hooks are always called **after** the `on_commit` ones.
I'm not sure this is the expected behaviour when we start mixing both 🤔 

Here are some test cases:
```
    def test_multiple_on_commit(self):
        with transaction.atomic():
            self.do(1)
            self.do(2)
            self.do(3)
        self.assertDone([1, 2, 3])

    def test_multiple_on_commit_robust(self):
        with transaction.atomic():
            self.do_robust(1)
            self.do_robust(2)
            self.do_robust(3)
        self.assertDone([1, 2, 3])

    def test_multiple_on_commit_robust_mix(self):
        with transaction.atomic():
            self.do(1)
            self.do_robust(2)
            self.do(3)
        self.assertDone([1, 2, 3])  # fails, because we get 1, 3, 2 currently
```
And I would expect them to be executed in the order they were declared.

One way to implement this, is to register all hooks to the same `self.run_on_commit` list, but instead of having 2-item tuples, append 3-element tuples in the list => `(sid, func, is_robust)` for instance. So you can iterate over them in defined order, and decide whether we try/catch or not depending on `is_robust`.
What do you think?

File: tests/transaction_hooks/tests.py
Comment (David-Wobrock) [inline tests/transaction_hooks/tests.py:42]: Subjective opinion: keep the same layout as `do`
=>
```suggestion
        Thing.objects.create(num=num)
        transaction.on_commit_robust(lambda: self.notify(num))
```

**Review by David-Wobrock — COMMENTED** [review]
Good work on the tests 👍 
I have one comment on the order in which we execute the handlers. See the discussion on the relevant code.

For the documentation:
1/ You can add a section in the release note
```
 What's new in Django 4.2
 ========================
 
+Performing actions that can fail after a transaction commit
+-----------------------------------------------------------
+
+The new :func:`~django.db.transaction.on_commit_robust` hook allows performing
+actions that can fail after a database transaction is successfully committed.
+This is useful to extend the behaviour of the existing
+:func:`~django.db.transaction.on_commit_robust`.
+
 Minor features
 --------------
```

2/ And mention it in `docs/topics/db/transactions.txt`
Maybe something like
```
 unhandled exception is raised in an :func:`atomic` block), your function will
 be discarded and never called.
 
+Django provides the :func:`on_commit_robust` function to register callback
+functions that can fail and should be executed after a transaction is
+successfully committed:
+
+.. function:: on_commit_robust(func, using=None)
+
+.. versionadded:: 4.2
+
 Savepoints
 ----------
```


But I know I'm often not hitting the right tone for documentation, a maintainer will be able to correct them later :)

File: django/db/backends/base/base.py
Comment (SirAbhi13) [reply django/db/backends/base/base.py:788]: > This implies that the on_commit_robust hooks are always called after the on_commit ones.
> I'm not sure this is the expected behaviour when we start mixing both 🤔

Yes, this is not the expected behaviour,

> One way to implement this, is to register all hooks to the same self.run_on_commit list, but instead of having 2-item tuples, append 3-element tuples in the list => (sid, func, is_robust) for instance. So you can iterate over them in defined order, and decide whether we try/catch or not depending on is_robust.
> What do you think?
Yes, I can implement your suggestion. That will be a nice solution and the functions will be executed in order. I will commit the changes soon. Thank you for the review!

File: tests/transaction_hooks/tests.py
Comment (SirAbhi13) [reply tests/transaction_hooks/tests.py:42]: Noted 👍

**Comment by SirAbhi13** [comment]
Made the suggested changes. Fixed the wrong behaviour of on_commit_robust()

File: django/db/backends/base/base.py
Comment (David-Wobrock) [inline django/db/backends/base/base.py:30]:
```suggestion
logger = logging.getLogger("django.db.backends")
```
is probably sufficient 🤔

File: docs/releases/4.2.txt
Comment (David-Wobrock) [inline docs/releases/4.2.txt:35]:
```suggestion
:func:`~django.db.transaction.on_commit`.
```

**Review by David-Wobrock — COMMENTED** [review]
Awesome 👍  Only two tiny adjustments and the patch looks good to me 🚀

File: django/db/backends/base/base.py
Comment (SirAbhi13) [reply django/db/backends/base/base.py:30]: I will make the change. I just used `django.db.backends.base` since the base.py file was in base module.

Comment (David-Wobrock) [reply django/db/backends/base/base.py:30]: I'm mostly mimicking existing behaviour. I can't find any logger with the `.base` module, but `django/db/backends/utils.py` has a plain `logger = logging.getLogger("django.db.backends")`. That's why I think we don't need the `base` :)

File: tests/transaction_hooks/tests.py
Comment (David-Wobrock) [inline tests/transaction_hooks/tests.py:59]: These will need to be adapted.

**Review by David-Wobrock — APPROVED** [review]
Looks good to me 👍

File: docs/releases/4.2.txt
Comment (felixxm) [inline docs/releases/4.2.txt:35]: This is not a major feature, please move release notes to the `Models`.

**Review by felixxm — COMMENTED** [review]
@SirAbhi13 Thanks for this patch :+1:

The execution in `captureOnCommitCallbacks()` should also be done in a `robust` fashion when the callback is register as `robust` (see [comment](https://code.djangoproject.com/ticket/33616#comment:7)).

Moreover, most of the new tests would be redundant if we change to `on_commit(..., robust=False)`.

@jarshwah Do you want to take a look?

**Comment by SirAbhi13** [comment]
@felixxm Should I move forward with the changes you suggested as I still haven't received a review from @jarshwah regarding if this method of adding this feature is correct or not. 

Please tell your thoughts!

**Comment by felixxm** [comment]
> @felixxm Should I move forward with the changes you suggested as I still haven't received a review from @jarshwah regarding if this method of adding this feature is correct or not.
> 
> Please tell your thoughts!

Please implement.

File: django/db/transaction.py
Comment (felixxm) [inline django/db/transaction.py:128]: We cannot change the order of kwargs:
```suggestion
def on_commit(func, using=None, is_robust=False):
```

File: django/test/testcases.py
Comment (felixxm) [inline django/test/testcases.py:1532]: I'd revert the logic i.e. check `execute` first:
```suggestion
                    if execute:
                        if is_robust:
                            try:
                                callback()
                            except Exception as e:
                                logger.error(
                                    f"Error calling {callback.__qualname__} on_commit() "
                                    f"during transaction (%s).",
                                    e,
                                    exc_info=True,
                                )
                        else:
                            callback()
```

File: django/db/backends/base/base.py
Comment (felixxm) [inline django/db/backends/base/base.py:731]:
```suggestion
    def on_commit(self, func, is_robust=False):
```
Add default value for better backward compatibility.

File: tests/transaction_hooks/tests.py
Comment (felixxm) [inline tests/transaction_hooks/tests.py:121]: `()` are unnecessary
```suggestion
                raise ForcedError
```

File: docs/topics/db/transactions.txt
Comment (felixxm) [inline docs/topics/db/transactions.txt:334]:
```suggestion
.. versionchanged:: 4.2

    The ``is_robust`` argument was added.
```

File: docs/topics/db/transactions.txt
Comment (felixxm) [inline docs/topics/db/transactions.txt:330]:
```suggestion
the current function throws an exception.
```

Comment (felixxm) [reply docs/topics/db/transactions.txt:330]: Wrap docs at 79 chars.

Comment (felixxm) [reply docs/topics/db/transactions.txt:330]: We should add that exceptions are logged (an where).

**Review by felixxm — COMMENTED** [review]
@SirAbhi13 Thanks :+1: I left more comments.

File: django/db/backends/base/base.py
Comment (SirAbhi13) [reply django/db/backends/base/base.py:731]: I didn't add a default value here since the function would receive a value from the `django/db/transaction.py`  on_commit function and a default value is added there. I will make the change if you deem it necessary in base.py file too. 👍

Comment (felixxm) [reply django/db/backends/base/base.py:731]: `on_commit` can be used directly by 3rd-party database backends, that's why we should add a default value to `is_robust.`

**Review by David-Wobrock — APPROVED** [review]

File: django/db/backends/base/base.py
Comment (felixxm) [reply django/db/backends/base/base.py]: `django.db.backends` logger is [documented](https://docs.djangoproject.com/en/4.1/ref/logging/#django-db-backends) as the SQL statements logger, so I will revert to `django.db.backends.base`.

**Review by felixxm — COMMENTED** [review]
@SirAbhi13 Thanks for updates :+1: 

I removed redundant tests, added the test method for robust callbacks in `captureOnCommitCallbacks()`, and pushed small edits to docs.

**Comment by adamchainz** [comment]
Slightly late bikeshedding: can we call the argument `robust` instead of `is_robust`? We don't generally use an `is_` prefix for boolean flags throughout Django.

File: docs/topics/db/transactions.txt
Comment (adamchainz) [inline docs/topics/db/transactions.txt:331]: `docs/ref/logging.txt` contains docs for all logging that Django does, afaik, so it would be good to add a section there too.

**Comment by felixxm** [comment]
> Slightly late bikeshedding: can we call the argument `robust` instead of `is_robust`? We don't generally use an `is_` prefix for boolean flags throughout Django.

I was thinking about this, renamed.

File: docs/topics/db/transactions.txt
Comment (felixxm) [reply docs/topics/db/transactions.txt:331]: > `docs/ref/logging.txt` contains docs for all logging that Django does, afaik, ...

We have several undocumented loggers i.e. `django.utils.autoreload`, `django.dispatch`, `django.db.models`, and `django.contrib.gis`.

> ... so it would be good to add a section there too.

Only folks that use `robust=True` will use this logger, so, IMO it's enough to mention it here.

**Pull request merged**
